src/fsharp/service/FSharpCheckerResults.fs

```diff
@@ -48,18 +48,70 @@ open FSharp.Compiler.Text.Position
 open FSharp.Compiler.Text.Range
 open FSharp.Compiler.TypedTree
 open FSharp.Compiler.TypedTreeOps
+open FSharp.Compiler.AbstractIL
 open System.Reflection.PortableExecutable
 
 open Internal.Utilities
 open Internal.Utilities.Collections
+open Internal.Utilities.Library
 open FSharp.Compiler.AbstractIL.ILBinaryReader
 
 type FSharpUnresolvedReferencesSet = FSharpUnresolvedReferencesSet of UnresolvedAssemblyReference list
 
-[<RequireQualifiedAccess;NoComparison>]
+[<Sealed>]
+type internal DelayedILModuleReader =
+
+    val private name : string
+    val private gate : obj
+    val mutable private getStream : (CancellationToken -> Stream option)
+    val mutable private result : ILModuleReader
+
+    new (name, getStream) = { name = name; gate = obj(); getStream = getStream; result = Unchecked.defaultof<_> }
```

Comment:
Any reason not to just make these `let-bound` values and assign `name` and `getStream` via a primary constructor?

Reply:
Yes, being explicit like this ensures I know exactly what `DelayedILModuleReader` is going to hold.

Reply:
@TIHan is setting the trend for how allocation sensitive F# code should look like :)

@TIHan, once you've got a good grasp and the style settled, it would be worth contributing to the F# style guidelines with your recommendations on the matter.